shared/src/business/useCases/docketEntry/fileCourtIssuedDocketEntryInteractor.js

```diff
@@ -31,128 +31,163 @@ exports.fileCourtIssuedDocketEntryInteractor = async (
+          documentTitle: documentMeta.generatedDocumentTitle,
+          documentType: documentMeta.documentType,
+          draftOrderState: null,
+          editState: JSON.stringify(documentMeta),
+          eventCode: documentMeta.eventCode,
+          filingDate: documentMeta.filingDate,
+          freeText: documentMeta.freeText,
+          isDraft: false,
+          isFileAttached: true,
+          isOnDocketRecord: true,
+          judge: documentMeta.judge,
+          numberOfPages,
+          scenario: documentMeta.scenario,
+          serviceStamp: documentMeta.serviceStamp,
+          trialLocation: documentMeta.trialLocation,
+          userId: user.userId,
+        },
+        { applicationContext },
+      );
+
+      const workItem = new WorkItem(
+        {
+          assigneeId: null,
+          assigneeName: null,
+          associatedJudge: caseToUpdate.associatedJudge,
+          caseStatus: caseToUpdate.status,
+          caseTitle: Case.getCaseTitle(caseEntity.caseCaption),
+          docketEntry: {
+            ...docketEntryEntity.toRawObject(),
+            createdAt: docketEntryEntity.createdAt,
+          },
+          docketNumber: caseToUpdate.docketNumber,
+          docketNumberWithSuffix: caseToUpdate.docketNumberWithSuffix,
+          hideFromPendingMessages: true,
+          inProgress: true,
+          section: DOCKET_SECTION,
+          sentBy: user.name,
+          sentByUserId: user.userId,
+        },
+        { applicationContext },
+      );
+
+      if (isUnservable) {
+        workItem.setAsCompleted({ message: 'completed', user });
+      }
+
+      docketEntryEntity.setWorkItem(workItem);
+
+      const isDocketEntryAlreadyOnCase = !!caseEntity.getDocketEntryById({
+        docketEntryId,
+      });
+
+      if (!isDocketEntryAlreadyOnCase) {
+        caseEntity.addDocketEntry(docketEntryEntity);
+      } else {
+        caseEntity.updateDocketEntry(docketEntryEntity);
+      }
+
+      workItem.assignToUser({
+        assigneeId: user.userId,
+        assigneeName: user.name,
+        section: user.section,
+        sentBy: user.name,
+        sentBySection: user.section,
+        sentByUserId: user.userId,
+      });
+
+      const saveItems = [
+        applicationContext.getUseCaseHelpers().updateCaseAndAssociations({
+          applicationContext,
+          caseToUpdate: caseEntity,
+        }),
+      ];
+
+      const rawValidWorkItem = workItem.validate().toRawObject();
+      if (isUnservable) {
+        saveItems.push(
+          applicationContext.getPersistenceGateway().putWorkItemInUsersOutbox({
+            applicationContext,
+            section: user.section,
+            userId: user.userId,
+            workItem: rawValidWorkItem,
+          }),
+        );
+      } else {
+        saveItems.push(
+          applicationContext.getPersistenceGateway().saveWorkItem({
+            applicationContext,
+            workItem: rawValidWorkItem,
+          }),
+        );
+      }
```

Comment:
🌱 This block of code exhibits one of the few things that does make me think TypeScript would be a good addition. Not that it would necessarily be an overall good decision to implement, but I think this could be much clearer. We have these two gateways that both pertain to `WorkItem`s, but we just have to know that they are both there and what the consequences of the two different functions are. If we could have an `interface` or a `type` which would let us just call the function we need on the the `WorkItem` and not have to know about these separate gateways. Of course we could implement something that allows similar behavior, but probably not as cleanly.